File: backend/repositories/pickingRepository.js
// backend/repositories/pickingRepository.js
import { WMS_STATUS } from "../config/wmsConstants.js";
import Logger from "../utils/logger.js";

// ============================================================================
// READ OPERATIONS
// ============================================================================

/**
 * @param {import('mysql2/promise').Connection} connection
 * @returns {Promise<any>}
 */
export const getPendingItems = async (connection) => {
  const query = `
    SELECT
      pli.id,
      pli.picking_list_id,
      pli.product_id,
      pli.original_sku as sku,
      pli.quantity,
      pli.status,
      COALESCE(loc_picked.code, loc_suggested.code) as location_code,
      p.name as product_name,
      pl.original_invoice_id,
      pl.source,
      pl.order_date,
      pl.created_at,
      pl.customer_name,
      pl.marketplace_status,
      pl.location_purpose,
      pl.shop_name,
      COALESCE(sl.quantity, 0) as available_stock
    FROM picking_list_items pli
    JOIN picking_lists pl ON pli.picking_list_id = pl.id
    LEFT JOIN products p ON pli.product_id = p.id
    LEFT JOIN locations loc_suggested ON pli.suggested_location_id = loc_suggested.id
    LEFT JOIN locations loc_picked ON pli.picked_from_location_id = loc_picked.id
    -- Join stok berdasarkan lokasi yang disarankan
    LEFT JOIN stock_locations sl ON sl.location_id = pli.suggested_location_id AND sl.product_id = pli.product_id
    WHERE pl.status IN (?, ?)
      AND pl.is_active = 1
      AND pli.status IN (?, 'BACKORDER')
    ORDER BY pl.created_at DESC, location_code ASC
  `;

  const [rows] = await connection.query(query, [
    WMS_STATUS.PENDING,
    WMS_STATUS.VALIDATED,
    WMS_STATUS.PENDING,
  ]);
  return rows;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number} limit
 * @returns {Promise<any>}
 */
export const getHistoryItems = async (connection, limit = 1000) => {
  const query = `
    SELECT
      pl.id as picking_list_id, pl.original_invoice_id, pl.source, pl.status,
      pl.marketplace_status, pl.customer_name, pl.shop_name, pl.created_at, pl.order_date, pl.location_purpose,
      pli.id as item_id, pli.original_sku as sku, pli.quantity, pli.status as item_status,
      pli.return_condition, pli.return_notes,
      p.name as product_name
    FROM picking_lists pl
    JOIN picking_list_items pli ON pl.id = pli.picking_list_id
    LEFT JOIN products p ON pli.product_id = p.id
    WHERE pl.status NOT IN (?)
    ORDER BY pl.created_at DESC, pl.id DESC
    LIMIT ?
  `;

  const [rows] = await connection.query(query, [WMS_STATUS.PENDING, limit]);
  return rows;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} pickingListId
 * @returns {Promise<any>}
 */
export const getListDetails = async (connection, pickingListId) => {
  const query = `
    SELECT
      pli.id,
      pli.original_sku as sku,
      pli.quantity as qty,
      p.name,
      pli.status,
      pli.return_condition,
      pli.return_notes
    FROM picking_list_items pli
    JOIN products p ON pli.product_id = p.id
    WHERE pli.picking_list_id = ?
  `;
  const [items] = await connection.query(query, [pickingListId]);
  return items;
};

// Added original_sku so Service can log errors properly
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} itemIds
 * @returns {Promise<any>}
 */
export const getItemsByIds = async (connection, itemIds) => {
  if (itemIds.length === 0) return [];
  const [rows] = await connection.query(
    `SELECT id, picking_list_id, suggested_location_id, product_id, quantity, original_sku, status
      FROM picking_list_items WHERE id IN (?)`,
    [itemIds]
  );
  return rows;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const countPendingItems = async (connection, listId) => {
  const [rows] = await connection.query(
    `SELECT COUNT(*) as count FROM picking_list_items
      WHERE picking_list_id = ? AND status NOT IN ('VALIDATED', 'CANCEL', 'COMPLETED', 'RETURNED', 'OBSOLETE', 'ERROR')`,
    [listId]
  );
  return rows[0].count;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} invoiceId
 * @returns {Promise<any>}
 */
export const findActiveHeaderByInvoice = async (connection, invoiceId) => {
  const [rows] = await connection.query(
    "SELECT id, status FROM picking_lists WHERE original_invoice_id = ? AND is_active = 1",
    [invoiceId]
  );
  return rows.length > 0 ? rows[0] : null;
};

// Required for Safety Check in pickingDataService
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const getHeaderById = async (connection, listId) => {
  const [rows] = await connection.query(
    "SELECT id, original_invoice_id, status, is_active, marketplace_status, location_purpose FROM picking_lists WHERE id = ?",
    [listId]
  );
  return rows.length > 0 ? rows[0] : null;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const getItemsForComparison = async (connection, listId) => {
  const [rows] = await connection.query(
    "SELECT original_sku as sku, quantity as qty FROM picking_list_items WHERE picking_list_id = ?",
    [listId]
  );
  return rows;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const getExistingItemSkus = async (connection, listId) => {
  const [rows] = await connection.query(
    "SELECT original_sku FROM picking_list_items WHERE picking_list_id = ?",
    [listId]
  );
  return new Set(rows.map((i) => i.original_sku));
};

// ============================================================================
// WRITE OPERATIONS (UPDATE)
// ============================================================================

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const cancelHeader = async (connection, listId) => {
  return connection.query(
    `UPDATE picking_lists SET status = ?, is_active = NULL, updated_at = NOW() WHERE id = ?`,
    [WMS_STATUS.CANCEL, listId]
  );
};

// Update Header Status
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @param {any} status
 * @param {boolean} isActive
 * @returns {Promise<any>}
 */
export const updateHeaderStatus = async (connection, listId, status, isActive) => {
  return connection.query(`UPDATE picking_lists SET status = ?, is_active = ? WHERE id = ?`, [
    status,
    isActive,
    listId,
  ]);
};

// Update Items Status by List ID
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @param {any} status
 * @returns {Promise<any>}
 */
export const updateItemsStatusByListId = async (connection, listId, status) => {
  return connection.query(`UPDATE picking_list_items SET status = ? WHERE picking_list_id = ?`, [
    status,
    listId,
  ]);
};

// Archives header by renaming with _REV_ suffix
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const archiveHeader = async (connection, listId) => {
  return connection.query(
    `UPDATE picking_lists
      SET status = 'OBSOLETE',
        is_active = NULL,
        original_invoice_id = CONCAT(original_invoice_id, '_REV_', id),
        updated_at = NOW()
      WHERE id = ?`,
    [listId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const cancelItemsByListId = async (connection, listId) => {
  return connection.query(`UPDATE picking_list_items SET status = ? WHERE picking_list_id = ?`, [
    WMS_STATUS.CANCEL,
    listId,
  ]);
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} itemId
 * @param {number|string} locationId
 * @returns {Promise<any>}
 */
export const updateSuggestedLocation = async (connection, itemId, locationId) => {
  Logger.info("Updating suggested location", "PICKING_REPO", { itemId, locationId });
  return connection.query(`UPDATE picking_list_items SET suggested_location_id = ? WHERE id = ?`, [
    locationId,
    itemId,
  ]);
};

// PENTING: Fungsi ini mengupdate status sekaligus mengunci lokasi final
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} itemId
 * @param {number|string} locationId
 * @returns {Promise<any>}
 */
export const validateItem = async (connection, itemId, locationId) => {
  return connection.query(
    `UPDATE picking_list_items
      SET status = 'VALIDATED', picked_from_location_id = ?, confirmed_location_id = ?
      WHERE id = ?`,
    [locationId, locationId, itemId]
  );
};

// Opsional: Jika Anda butuh fungsi khusus untuk update status saja tanpa lokasi
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} itemId
 * @param {any} status
 * @returns {Promise<any>}
 */
export const updateItemStatus = async (connection, itemId, status) => {
  return connection.query(`UPDATE picking_list_items SET status = ? WHERE id = ?`, [
    status,
    itemId,
  ]);
};

// Marks specific item as RETURNED with condition notes
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} itemId
 * @param {any} condition
 * @param {any} notes
 * @returns {Promise<any>}
 */
export const markItemAsReturned = async (connection, itemId, condition, notes) => {
  return connection.query(
    `UPDATE picking_list_items
      SET status = 'RETURNED',
        return_condition = ?,
        return_notes = ?
      WHERE id = ?`,
    [condition, notes, itemId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @returns {Promise<any>}
 */
export const validateHeader = async (connection, listId) => {
  return connection.query(
    `UPDATE picking_lists SET status = 'VALIDATED', updated_at = NOW() WHERE id = ?`,
    [listId]
  );
};

// Updates Marketplace status, optionally syncing WMS status too
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} listId
 * @param {any} mpStatus
 * @param {any} wmsStatus
 * @returns {Promise<any>}
 */
export const updateMarketplaceStatus = async (connection, listId, mpStatus, wmsStatus = null) => {
  if (wmsStatus) {
    return connection.query(
      `UPDATE picking_lists SET marketplace_status = ?, status = ? WHERE id = ?`,
      [mpStatus, wmsStatus, listId]
    );
  }
  return connection.query(`UPDATE picking_lists SET marketplace_status = ? WHERE id = ?`, [
    mpStatus,
    listId,
  ]);
};

// Bulk Check Existing Invoices (ACTIVE ONLY)
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} invoiceIds
 * @returns {Promise<any>}
 */
export const getActiveHeadersByInvoiceIds = async (connection, invoiceIds) => {
  if (invoiceIds.length === 0) return [];
  const [rows] = await connection.query(
    `SELECT id, original_invoice_id, status, marketplace_status
      FROM picking_lists
      WHERE original_invoice_id IN (?) AND is_active = 1`,
    [invoiceIds]
  );
  return rows;
};

// Bulk Check Existing Invoices (ALL - Active & Inactive/Cancelled)
// Ini diperlukan untuk mendeteksi duplicate entry pada order yang sudah dicancel
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} invoiceIds
 * @returns {Promise<any>}
 */
export const getAllHeadersByInvoiceIds = async (connection, invoiceIds) => {
  if (invoiceIds.length === 0) return [];
  const [rows] = await connection.query(
    `SELECT id, original_invoice_id, status, marketplace_status, is_active
      FROM picking_lists
      WHERE original_invoice_id IN (?)`,
    [invoiceIds]
  );
  return rows;
};

// ============================================================================
// CREATE OPERATIONS
// ============================================================================

export const createHeader = async (
  connection,
  { userId, source, invoiceId, customer, orderDate, status, mpStatus, locationPurpose, shopName }
) => {
  const [res] = await connection.query(
    `INSERT INTO picking_lists (user_id, source, original_invoice_id, customer_name, order_date, status, marketplace_status, is_active, created_at, location_purpose, shop_name)
      VALUES (?, ?, ?, ?, ?, ?, ?, 1, NOW(), ?, ?)`,
    [userId, source, invoiceId, customer, orderDate, status, mpStatus, locationPurpose || "DISPLAY", shopName || null]
  );
  return res.insertId;
};

export const createItem = async (
  connection,
  { listId, productId, sku, qty, status, locationId = null }
) => {
  return connection.query(
    `INSERT INTO picking_list_items (picking_list_id, product_id, original_sku, quantity, status, suggested_location_id)
      VALUES (?, ?, ?, ?, ?, ?)`,
    [listId, productId, sku, qty, status, locationId]
  );
};

// Bulk Create Items
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {any} rows
 * @returns {Promise<any>}
 */
export const createItemsBulk = async (connection, rows) => {
  if (rows.length === 0) return;
  return connection.query(
    `INSERT INTO picking_list_items (picking_list_id, product_id, original_sku, price, quantity, status, suggested_location_id) VALUES ?`,
    [rows]
  );
};

File: backend/internal/repository/picking_repository.go
package repository

import (
	"context"
	"database/sql"
	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/model"
	"github.com/jmoiron/sqlx"
)

type PickingRepository interface {
	GetPendingItems(ctx context.Context) ([]dto.PendingPickingItemResponse, error)
	GetHistoryItems(ctx context.Context, limit int) ([]dto.HistoryPickingItemResponse, error)
	GetListDetails(ctx context.Context, pickingListID int) ([]dto.PickingListDetailResponse, error)

	GetHeaderByID(ctx context.Context, tx *sqlx.Tx, id int) (*model.PickingList, error)
	GetItemsByIDs(ctx context.Context, tx *sqlx.Tx, itemIDs []int) ([]model.PickingListItem, error)
	CountPendingItems(ctx context.Context, tx *sqlx.Tx, pickingListID int) (int, error)

	UpdateSuggestedLocation(ctx context.Context, tx *sqlx.Tx, itemID int, locationID *int) error
	UpdateItemStatus(ctx context.Context, tx *sqlx.Tx, itemID int, status string) error
	ValidateItem(ctx context.Context, tx *sqlx.Tx, itemID int, locationID int) error
	ValidateHeader(ctx context.Context, tx *sqlx.Tx, listID int) error

	VoidHeader(ctx context.Context, tx *sqlx.Tx, listID int) (int64, error)
	VoidItemsByListID(ctx context.Context, tx *sqlx.Tx, listID int) error
	
	GetItemsToRestock(ctx context.Context, tx *sqlx.Tx, listID int) ([]model.PickingListItem, error)
	GetUnfulfillableItems(ctx context.Context, tx *sqlx.Tx, listID int) ([]model.PickingListItem, error)
	GetPendingAndBackorderItems(ctx context.Context, tx *sqlx.Tx, listIDs []int) ([]model.PickingListItem, error)
}

type pickingRepository struct {
	db *sqlx.DB
}

func NewPickingRepository(db *sqlx.DB) PickingRepository {
	return &pickingRepository{db: db}
}

func (r *pickingRepository) GetPendingItems(ctx context.Context) ([]dto.PendingPickingItemResponse, error) {
	query := `
		SELECT
			pli.id,
			pli.picking_list_id,
			pli.product_id,
			pli.original_sku as sku,
			pli.quantity,
			pli.status,
			COALESCE(loc_picked.code, loc_suggested.code) as location_code,
			p.name as product_name,
			pl.original_invoice_id,
			pl.source,
			pl.order_date,
			pl.created_at,
			pl.customer_name,
			pl.marketplace_status,
			pl.location_purpose,
			pl.shop_name,
			COALESCE(sl.quantity, 0) as available_stock
		FROM picking_list_items pli
		JOIN picking_lists pl ON pli.picking_list_id = pl.id
		LEFT JOIN products p ON pli.product_id = p.id
		LEFT JOIN locations loc_suggested ON pli.suggested_location_id = loc_suggested.id
		LEFT JOIN locations loc_picked ON pli.picked_from_location_id = loc_picked.id
		LEFT JOIN stock_locations sl ON sl.location_id = pli.suggested_location_id AND sl.product_id = pli.product_id
		WHERE pl.status IN (?, ?)
		  AND pl.is_active = 1
		  AND pli.status IN (?, 'BACKORDER')
		ORDER BY pl.created_at DESC, location_code ASC
	`
	var items []dto.PendingPickingItemResponse
	err := r.db.SelectContext(ctx, &items, query, "PENDING", "VALIDATED", "PENDING")
	return items, err
}

func (r *pickingRepository) GetHistoryItems(ctx context.Context, limit int) ([]dto.HistoryPickingItemResponse, error) {
	query := `
		SELECT
			pl.id as picking_list_id, pl.original_invoice_id, pl.source, pl.status,
			pl.marketplace_status, pl.customer_name, pl.shop_name, pl.created_at, pl.order_date, pl.location_purpose,
			pli.id as item_id, pli.original_sku as sku, pli.quantity, pli.status as item_status,
			pli.return_condition, pli.return_notes,
			p.name as product_name
		FROM picking_lists pl
		JOIN picking_list_items pli ON pl.id = pli.picking_list_id
		LEFT JOIN products p ON pli.product_id = p.id
		WHERE pl.status NOT IN (?)
		ORDER BY pl.created_at DESC, pl.id DESC
		LIMIT ?
	`
	var items []dto.HistoryPickingItemResponse
	err := r.db.SelectContext(ctx, &items, query, "PENDING", limit)
	return items, err
}

func (r *pickingRepository) GetListDetails(ctx context.Context, pickingListID int) ([]dto.PickingListDetailResponse, error) {
	query := `
		SELECT
			pli.id,
			pli.original_sku as sku,
			pli.quantity as qty,
			p.name,
			pli.status,
			pli.return_condition,
			pli.return_notes
		FROM picking_list_items pli
		JOIN products p ON pli.product_id = p.id
		WHERE pli.picking_list_id = ?
	`
	var items []dto.PickingListDetailResponse
	err := r.db.SelectContext(ctx, &items, query, pickingListID)
	return items, err
}

func (r *pickingRepository) GetHeaderByID(ctx context.Context, tx *sqlx.Tx, id int) (*model.PickingList, error) {
	query := `SELECT * FROM picking_lists WHERE id = ? LIMIT 1`
	var header model.PickingList
	var err error
	if tx != nil {
		err = tx.GetContext(ctx, &header, query, id)
	} else {
		err = r.db.GetContext(ctx, &header, query, id)
	}
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &header, err
}

func (r *pickingRepository) GetItemsByIDs(ctx context.Context, tx *sqlx.Tx, itemIDs []int) ([]model.PickingListItem, error) {
	if len(itemIDs) == 0 {
		return []model.PickingListItem{}, nil
	}
	query, args, err := sqlx.In(`SELECT * FROM picking_list_items WHERE id IN (?)`, itemIDs)
	if err != nil {
		return nil, err
	}
	query = tx.Rebind(query)
	var items []model.PickingListItem
	err = tx.SelectContext(ctx, &items, query, args...)
	return items, err
}

func (r *pickingRepository) CountPendingItems(ctx context.Context, tx *sqlx.Tx, pickingListID int) (int, error) {
	query := `SELECT count(*) FROM picking_list_items WHERE picking_list_id = ? AND status IN ('PENDING', 'BACKORDER')`
	var count int
	err := tx.GetContext(ctx, &count, query, pickingListID)
	return count, err
}

func (r *pickingRepository) UpdateSuggestedLocation(ctx context.Context, tx *sqlx.Tx, itemID int, locationID *int) error {
	query := `UPDATE picking_list_items SET suggested_location_id = ?, updated_at = NOW() WHERE id = ?`
	_, err := tx.ExecContext(ctx, query, locationID, itemID)
	return err
}

func (r *pickingRepository) UpdateItemStatus(ctx context.Context, tx *sqlx.Tx, itemID int, status string) error {
	query := `UPDATE picking_list_items SET status = ?, updated_at = NOW() WHERE id = ?`
	_, err := tx.ExecContext(ctx, query, status, itemID)
	return err
}

func (r *pickingRepository) ValidateItem(ctx context.Context, tx *sqlx.Tx, itemID int, locationID int) error {
	query := `UPDATE picking_list_items SET status = 'VALIDATED', picked_from_location_id = ?, updated_at = NOW() WHERE id = ?`
	_, err := tx.ExecContext(ctx, query, locationID, itemID)
	return err
}

func (r *pickingRepository) ValidateHeader(ctx context.Context, tx *sqlx.Tx, listID int) error {
	query := `UPDATE picking_lists SET status = 'VALIDATED', updated_at = NOW() WHERE id = ?`
	_, err := tx.ExecContext(ctx, query, listID)
	return err
}

func (r *pickingRepository) VoidHeader(ctx context.Context, tx *sqlx.Tx, listID int) (int64, error) {
	query := `UPDATE picking_lists SET status = 'VOID', is_active = 0, updated_at = NOW() WHERE id = ? AND status != 'VOID'`
	res, err := tx.ExecContext(ctx, query, listID)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

func (r *pickingRepository) VoidItemsByListID(ctx context.Context, tx *sqlx.Tx, listID int) error {
	query := `UPDATE picking_list_items SET status = 'VOID', updated_at = NOW() WHERE picking_list_id = ?`
	_, err := tx.ExecContext(ctx, query, listID)
	return err
}

func (r *pickingRepository) GetItemsToRestock(ctx context.Context, tx *sqlx.Tx, listID int) ([]model.PickingListItem, error) {
	query := `
		SELECT * FROM picking_list_items
		WHERE picking_list_id = ? AND status = 'VALIDATED' AND picked_from_location_id IS NOT NULL
	`
	var items []model.PickingListItem
	err := tx.SelectContext(ctx, &items, query, listID)
	return items, err
}

func (r *pickingRepository) GetUnfulfillableItems(ctx context.Context, tx *sqlx.Tx, listID int) ([]model.PickingListItem, error) {
	query := `
		SELECT pli.* FROM picking_list_items pli
		WHERE pli.picking_list_id = ?
		  AND (pli.status = 'BACKORDER' OR (pli.status = 'PENDING' AND pli.suggested_location_id IS NULL))
	`
	var items []model.PickingListItem
	err := tx.SelectContext(ctx, &items, query, listID)
	return items, err
}

func (r *pickingRepository) GetPendingAndBackorderItems(ctx context.Context, tx *sqlx.Tx, listIDs []int) ([]model.PickingListItem, error) {
	if len(listIDs) == 0 {
		return []model.PickingListItem{}, nil
	}
	query, args, err := sqlx.In(`
		SELECT pli.* FROM picking_list_items pli
		JOIN picking_lists pl ON pli.picking_list_id = pl.id
		WHERE pli.picking_list_id IN (?)
		  AND pli.status IN ('PENDING', 'BACKORDER')
		  AND pl.status IN ('PENDING', 'VALIDATED')
		  AND pl.is_active = 1
	`, listIDs)
	if err != nil {
		return nil, err
	}
	query = tx.Rebind(query)
	var items []model.PickingListItem
	err = tx.SelectContext(ctx, &items, query, args...)
	return items, err
}
